Rebase MediaArea patches on upstream zlib 1.3.1

cjee21 opened the pull request:
Pick and port some patches from upstream to eliminate MSVC compile-time warnings and reduce VS code analysis warnings.

JeromeMartinez committed:
BCB project files

JeromeMartinez committed:
Use DLL runtime library everywhere

JeromeMartinez committed:
Support of 32-bit CPU without SSE

Since Visual Studio 2010, default changed from IA32 to SSE2.
We keep same behavior everywhere (IA32 for all versions of Visual Studio)

JeromeMartinez committed:
MultiProcessorCompilation flag

JeromeMartinez committed:
Visual Studio DebugInformationFormat set to none in release mode

JeromeMartinez committed:
Removing ZLIB_WINAPI as it does not permit static linking everywhere

Co-Authored-By: cjee21 <77721854+cjee21@users.noreply.github.com>

g-maxime committed:
Use llvm based bcc32c compiler

Signed-off-by: Maxime Gervais <gervais.maxime@gmail.com>

g-maxime committed:
Add UWP project file for MSVC2022

Signed-off-by: Maxime Gervais <gervais.maxime@gmail.com>

cjee21 commented:
By the way, upstream has some ARM-related patches as well.

JeromeMartinez commented:
It is nice from you but I would prefer to rebase from upstream master rather than cherry picking.
But there are 23 commits to check and/or rework :( (some of them are useless e.g. adding vc17 and ARM64, but we need to be careful to miss nothing e.g. UWP).

Would you be interested in trying to rebase the MediaArea patches?

cjee21 commented:
> It is nice from you but I would prefer to rebase from upstream master rather than cherry picking. 

That's what I think is better too but I see many commits and conflicts so I resort to cherry picking.

> Would you be interested in trying to rebase the MediaArea patches?

I can try... since MediaArea's commits are mostly (if not all) touching only project files and not the codes, I should be able to do it I guess.

JeromeMartinez commented:
> I can try...

Thank you!

> since MediaArea's commits are mostly (if not all) touching only project files and not the codes, I should be able to do it I guess.

They are definitely only on project files, only tweaks for fitting some compilation needs on our side.

JeromeMartinez commented:
BTW, when possible please keep the name and date of the commits even if modified so it is easier to track when we did it in the past, maybe with an additional comment if there is something to add about the patch.

cjee21 committed:
Use DLL runtime library everywhere (VC17)

cjee21 committed:
Support of 32-bit CPU without SSE (VC17)

Since Visual Studio 2010, default changed from IA32 to SSE2.
We keep same behavior everywhere (IA32 for all versions of Visual Studio)

cjee21 committed:
MultiProcessorCompilation flag (VC17)

cjee21 committed:
Visual Studio DebugInformationFormat set to none in release mode (VC17)

cjee21 committed:
Enable GS for vc17

cjee21 committed:
+ Windows: Add ControlFlowGuard for Release

cjee21 committed:
+ Windows: Enable EH Continuation Metadata

cjee21 committed:
ARM64: Enable Signed Returns / Pointer Authentication

cjee21 committed:
Add ARM64X build for MSVC2022

cjee21 commented:
MediaInfoLib (Win32, x64, ARM64, ARM64X) builds fine with this branch. Let me know if I missed anything.

JeromeMartinez commented:
Conflicts :).
Not sure about the best method for that, I think I'll sync this master with upstream then you push the additional commits and I merge this PR as a package containing all the MediaArea updates.

cjee21 commented:
> Conflicts :). Not sure about the best method for that, I think I'll sync this master with upstream then you push the additional commits and I merge this PR as a package containing all the MediaArea updates.

master has to match upstream for this PR to work.

cjee21 commented:
Just realized zlib upstream defaults to develop branch. Switched to master as base.

cjee21 commented:
Need modify the CI / GitHub actions?

JeromeMartinez commented:
> Need modify the CI / GitHub actions?

Looks like that yes (IIRC we fixed that on our repos too, lot of GH things were deprecated on last January 1st).

cjee21 commented:
Note that vc15 and vc16 project files are not available upstream and are now no longer present in this repo. If they are still needed, they have to be re-created. Upstream has some significant changes so the old project files are not expected to work.

Previously they were added by commits https://github.com/MediaArea/zlib/commit/c355645712b2bde1a607f7558401cdb015cc1854 and https://github.com/MediaArea/zlib/commit/de09e5f9f767cd56f05329c4a73ab8d52eb6a664 and their latest state before the rebase on upstream can be found [here](https://github.com/MediaArea/zlib/tree/a72ef048cd2c61f7b519ecedbe8e5ff3501652da/contrib/vstudio).
